File: internal/web/update.go
// MASS updates itself by running its own installer over its own install: the
// daemon notices a newer release, downloads that release's mass-setup, and
// hands it the recorded install directory with --relaunch. The setup waits for
// this process to exit, stages the new build, and starts the app again.
//
// Finding a release and installing one both live in mass-sdk/selfupdate. What
// stays here is what only MASS knows: the fleet gate (a new build can strand
// connected workers) and the operator-managed daemon that won't restart itself.
package web

import (
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"

	"github.com/Masterminds/semver/v3"
	"github.com/chinese-room-solutions/mass-sdk/registry"
	"github.com/chinese-room-solutions/mass-sdk/selfupdate"
)

// RunUpdateChecks keeps the daemon's answer to "is there a newer MASS?" fresh:
// one check now, then another every selfupdate.DefaultInterval until ctx ends.
// A single check at startup is not enough — the dashboard can render before the
// first answer lands, and a daemon left running outlives releases published
// after it started.
func (h *Handler) RunUpdateChecks(ctx context.Context) {
	h.update.Run(ctx)
}

// errOperatorManaged refuses to restart a daemon nobody asked to be restarted.
// An on-demand daemon belongs to the GUI/CLI that spawned it and comes back by
// itself; a `mass serve` is a fleet hub someone runs deliberately, and taking it
// down from under its workers is not something a dashboard click gets to do.
// The SDK's own refusals (not installer-placed, needs elevation) come back out
// of Applier.Apply the same way.
var errOperatorManaged = selfupdate.Refuse(
	"this MASS runs as an operator-managed server, so it won't restart itself — " +
		"stop it and run the installer, or update it from the desktop app")

// errIncompatibleFleet is the fleet gate's refusal: connected workers the
// candidate build would strand. Count carries how many, Names a sample for the
// message. It is only advisory in the sense that {"force": true} overrides it.
type errIncompatibleFleet struct {
	Count int
	Names []string
	Tag   string
}

func (e *errIncompatibleFleet) Error() string {
	msg := fmt.Sprintf("%s would strand %s", e.Tag, workersPhrase(e.Count))
	if len(e.Names) > 0 {
		msg += " (" + joinNames(e.Names) + ")"
	}
	return msg + " — a stranded worker exits and stays down until it is upgraded. " +
		"Upgrade the workers first, or send {\"force\": true} to update anyway."
}

func workersPhrase(n int) string {
	if n == 1 {
		return "1 connected worker"
	}
	return fmt.Sprintf("%d connected workers", n)
}

func joinNames(names []string) string {
	out := ""
	for i, n := range names {
		if i > 0 {
			out += ", "
		}
		out += n
	}
	return out
}

// applyUpdate installs tag over this install — MASS's own refusal first, then
// the SDK's.
func (h *Handler) applyUpdate(ctx context.Context, tag string) error {
	// Same guardrail ensureDaemon applies when a skewed client meets a
	// non-on-demand daemon (cmd/mass/launch.go).
	if !h.onDemand {
		return errOperatorManaged
	}
	return h.applier.Apply(ctx, tag)
}

// --- The fleet pre-flight gate ---

// UpdateFleetGate is what the candidate-aware compat check found: how many
// connected workers the registry index says the candidate MASS build would
// reject at Register, and a sample of their names.
type UpdateFleetGate struct {
	Incompatible int      `json:"incompatible"`
	Names        []string `json:"names,omitempty"`
}

// fleetGateNameCap bounds the names carried in the response and the message —
// enough to recognise the fleet, not a full listing.
const fleetGateNameCap = 5

// updateFleetGate counts the connected workers whose registry-index row admits
// this MASS build but not candidate. It is the load-bearing half of the update
// flow: a version-skewed worker is rejected with FAILED_PRECONDITION at
// Register, which the worker treats as fatal — it exits and STAYS DOWN until
// somebody upgrades it (internal/worker/hub.go). So an unchecked MASS upgrade
// can silently empty the fleet.
//
// Cache-only, never network — the same posture as CheckWorkerCompat, whose
// verdict this predicts. Every inconclusive input (no cached index, no row for
// the worker's version, an unparseable range, a dev/pre-release version) counts
// as compatible, because that is exactly what Register would do with it:
// accept-with-warning. The gate must never claim a breakage the hub wouldn't
// enforce.
func (h *Handler) updateFleetGate(candidate string) UpdateFleetGate {
	idx, err := h.cachedIndex()
	if err != nil || idx == nil {
		h.logger.Debug().Err(err).Msg("no cached registry index for the update fleet gate")
		return UpdateFleetGate{}
	}
	return fleetGate(idx, h.fleetPairings(), candidate)
}

// fleetGate is updateFleetGate's verdict over an explicit index and fleet.
func fleetGate(idx *registry.Index, workers []workerPairing, candidate string) UpdateFleetGate {
	gate := UpdateFleetGate{}
	for _, w := range workers {
		if !workerRejectsMass(idx, w, candidate) {
			continue
		}
		gate.Incompatible++
		if len(gate.Names) < fleetGateNameCap {
			gate.Names = append(gate.Names, w.RuntimeName+" "+w.Version)
		}
	}
	return gate
}

// workerRejectsMass reports whether the index says w's version pairs with no
// MASS in candidate's range. Mirrors checkWorkerIndexCompat's row semantics:
// one admitting row is enough, and an inconclusive row admits.
func workerRejectsMass(idx *registry.Index, w workerPairing, candidate string) bool {
	wv, err := semver.NewVersion(w.Version)
	if err != nil {
		return false
	}
	rows := workerVersionRows(idx, w.RuntimeName, wv)
	if len(rows) == 0 {
		return false
	}
	for _, row := range rows {
		if ok, reason := admits(row.Mass, candidate); ok || reason != "" {
			return false
		}
	}
	return true
}

// --- HTTP surface ---

// UpdateCheckResponse is what both /api/update/check verbs answer: the running
// build, the newer tag (empty when there is none), when that answer was taken,
// why the last check failed, and the fleet the candidate would strand.
type UpdateCheckResponse struct {
	selfupdate.Status
	UpdateFleetGate
}

// handleUpdateCheck reports the last known answer and what applying it would
// cost the fleet. It reads stored state and the on-disk index cache — no
// network, so a dashboard may call it on every load.
func (h *Handler) handleUpdateCheck(w http.ResponseWriter, _ *http.Request) {
	h.writeUpdateStatus(w, h.update.Status())
}

// handleUpdateCheckNow asks the release repository right now — the operator
// pressing "Check for updates", or a `mass update` that must not report a
// never-taken answer. A check that can't reach the repository is not an HTTP
// error: the operator needs the sentence, which the status carries, and a
// status code would read as a broken daemon rather than an unreachable GitHub.
func (h *Handler) handleUpdateCheckNow(w http.ResponseWriter, r *http.Request) {
	st, err := h.update.Check(r.Context())
	if err != nil {
		h.logger.Debug().Err(err).Str("url", h.update.BaseURL).Msg("checking for a newer MASS release")
	}
	h.writeUpdateStatus(w, st)
}

// writeUpdateStatus answers with st, priced against the connected fleet.
func (h *Handler) writeUpdateStatus(w http.ResponseWriter, st selfupdate.Status) {
	resp := UpdateCheckResponse{Status: st}
	if st.Available != "" {
		resp.UpdateFleetGate = h.updateFleetGate(st.Available)
	}
	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(resp); err != nil {
		h.logger.Debug().Err(err).Msg("writing the update check response")
	}
}

// updateApplyRequest is POST /api/update/apply's optional body. force
// overrides the fleet gate — safe by default, but the operator who has read the
// warning can still choose.
type updateApplyRequest struct {
	Force bool `json:"force"`
}

// handleUpdateApply installs the release the check found. It does the parts
// that can fail — the gates, resolving the install, downloading and verifying
// the installer — inside the request, so a caller that gets a 200 knows the
// installer is running. Only then does it retire the daemon, from a goroutine:
// the shutdown drains this very request.
func (h *Handler) handleUpdateApply(w http.ResponseWriter, r *http.Request) {
	tag := h.update.Available()
	if tag == "" {
		http.Error(w, "no MASS update is available", http.StatusConflict)
		return
	}
	var req updateApplyRequest
	if r.Body != nil {
		// An empty or malformed body is simply "no force" — the flag is the only
		// thing in it, and a caller that sends nothing means the default.
		_ = json.NewDecoder(r.Body).Decode(&req)
	}
	if !req.Force {
		if gate := h.updateFleetGate(tag); gate.Incompatible > 0 {
			http.Error(w, (&errIncompatibleFleet{
				Count: gate.Incompatible, Names: gate.Names, Tag: tag,
			}).Error(), http.StatusConflict)
			return
		}
	}

	h.logger.Info().Str("tag", tag).Bool("force", req.Force).Msg("applying a MASS update")
	if err := h.applyUpdate(r.Context(), tag); err != nil {
		// A refusal is the operator's situation rather than a fault, so it
		// answers with the sentence to act on and a 409.
		status := http.StatusInternalServerError
		if selfupdate.IsRefusal(err) {
			status = http.StatusConflict
		} else {
			h.logger.Warn().Err(err).Msg("applying the MASS update")
		}
		http.Error(w, err.Error(), status)
		return
	}

	w.Header().Set("Content-Type", "application/json")
	if err := json.NewEncoder(w).Encode(map[string]string{"status": "updating", "version": tag}); err != nil {
		h.logger.Debug().Err(err).Msg("writing the update apply response")
	}

	// Tell the attached window before retiring. Without this it would just
	// reconnect to whatever answers on the same port next — which is the
	// relaunched build's own daemon — and the user would be left with two MASS
	// windows: this one re-attached, and the new instance's. With no window
	// attached (a headless serve, a browser) there is nobody to tell, and the
	// browser tab simply reloads against the new daemon.
	h.gui.send(guiEvent{name: GUIEventUpdateRestarting, data: tag})

	// The installer is waiting for this process's files to be replaceable, so
	// the daemon has to go. The window is quitting on the event above; the grace
	// gives that event its trip down the SSE stream before Shutdown closes it.
	h.shutdownMu.Lock()
	fn := h.shutdownFn
	h.shutdownMu.Unlock()
	if fn != nil {
		go func() {
			time.Sleep(updateShutdownGrace)
			fn()
		}()
	}
}

// updateShutdownGrace is the pause between telling the window an update is
// coming and tearing the server down. Only long enough for that event to be
// written and read — the window's own notice-then-quit grace runs in parallel,
// and the installer is already waiting on both processes to exit.
const updateShutdownGrace = 250 * time.Millisecond

File: internal/web/templates/scripts/update.js
(function() {
  // Self-update: the Settings > About Update row plus a one-per-load
  // announcement. The row is always in the page and always starts empty — this
  // script hydrates it from api/update/check, so an answer the daemon hadn't
  // got yet when the page rendered no longer hides the whole affordance. The
  // announcement is a toast rather than a modal — nothing is broken — with the
  // install offered right in it; the About row is what remains after it clears.
  //
  // The fleet gate is the MASS-specific part. When the daemon reports connected
  // workers the new build would strand, the toast says so and the install goes
  // through an explicit confirm before it re-sends with force — a stranded
  // worker exits at Register and stays down until it is upgraded.
  // The window process calls this over the webview bridge when the daemon says
  // an update is being installed: MASS is about to be replaced and the window
  // closed behind this notice. Sticky (duration Infinity) — the window goes away
  // before any timeout would matter, and a notice that vanished first would
  // leave the closing window unexplained. Defined before the row guard: the
  // update may have been applied from the CLI, or from another browser tab.
  // Raised at most once: in the desktop app the clicking page shows it on the
  // 200 and the daemon pushes the same event to the window process a moment
  // later, and two identical sticky toasts would stack.
  var restartingShown = false;
  window.massUpdateRestarting = function(tag) {
    if (restartingShown) return;
    restartingShown = true;
    window.massToast('Updating MASS to ' + tag + ' — restarting…', {duration: Infinity});
  };

  var statusEl = document.getElementById('mass-update-status');
  if (!statusEl) return;
  var checkBtn = document.getElementById('mass-update-check-btn');
  var installBtn = document.getElementById('mass-update-btn');
  var warnBox = document.getElementById('mass-update-warning');
  var warnText = document.getElementById('mass-update-warning-text');

  // What the last answer said, so a failed live check can be shown over it
  // rather than throwing the known tag away.
  var last = {};
  var tag = '';
  var incompatible = 0;
  var announced = false;

  function workersPhrase(n) {
    return n === 1 ? '1 connected worker' : n + ' connected workers';
  }

  // checkedLabel is the relative age of an answer. A never-taken one carries
  // Go's zero time, which parses to a negative epoch — hence the > 0 guard.
  function checkedLabel(ts) {
    var t = Date.parse(ts || '');
    if (!(t > 0)) return '';
    var s = (Date.now() - t) / 1000;
    if (s < 90) return 'checked just now';
    if (s < 3600) return 'checked ' + Math.round(s / 60) + 'm ago';
    if (s < 86400) return 'checked ' + Math.round(s / 3600) + 'h ago';
    return 'checked ' + Math.round(s / 86400) + 'd ago';
  }

  function render(st) {
    last = st;
    tag = st.available || '';
    incompatible = st.incompatible || 0;
    installBtn.style.display = tag ? '' : 'none';
    warnBox.style.display = tag && incompatible > 0 ? '' : 'none';
    if (incompatible > 0) {
      warnText.textContent = workersPhrase(incompatible) + ' incompatible';
    }
    if (st.error) {
      // An unreachable release host must not read as "up to date": say so, and
      // say what went wrong.
      statusEl.style.color = 'var(--mass-danger)';
      statusEl.textContent = "Couldn't check for updates — " + st.error;
      return;
    }
    statusEl.style.color = 'var(--mass-text-muted)';
    if (!tag) {
      var when = checkedLabel(st.checked_at);
      statusEl.textContent = when ? 'Up to date — ' + when : 'Up to date';
      return;
    }
    statusEl.textContent = tag + ' available';
    announce();
  }

  function announce() {
    if (announced) return;
    announced = true;
    // 15s rather than the default: the toast carries the action, and the
    // default is short enough to vanish mid-reach.
    var msg = 'MASS ' + tag + ' is available';
    if (incompatible > 0) msg += ' — ' + workersPhrase(incompatible) + ' would be stranded';
    window.massToast(msg, {
      duration: 15000,
      variant: incompatible > 0 ? 'warning' : undefined,
      action: {label: 'Install', onClick: apply},
    });
  }

  // check asks the daemon to go to the network now. The daemon answers 200 with
  // an error field when it couldn't reach the release host, so only a broken
  // request lands in the rejection path.
  function check() {
    checkBtn.loading = true;
    checkBtn.disabled = true;
    return fetch('api/update/check', {method: 'POST'}).then(function(r) {
      if (!r.ok) return Promise.reject(new Error('HTTP ' + r.status));
      return r.json();
    }).then(render, function() {
      last.error = "MASS isn't responding.";
      render(last);
    }).then(function() {
      checkBtn.loading = false;
      checkBtn.disabled = false;
    });
  }

  // hydrate reads the daemon's cached answer — no network on its side. When it
  // has never taken one (the daemon started moments ago, which is exactly the
  // race this page used to lose), ask for a live one instead of showing nothing.
  function hydrate() {
    fetch('api/update/check').then(function(r) {
      if (!r.ok) return Promise.reject(new Error('HTTP ' + r.status));
      return r.json();
    }).then(function(st) {
      if (!(Date.parse(st.checked_at || '') > 0)) {
        check();
        return;
      }
      render(st);
    }, function() {
      statusEl.textContent = '';
    });
  }

  function post(force) {
    installBtn.disabled = true;
    return fetch('api/update/apply', {
      method: 'POST',
      headers: {'Content-Type': 'application/json'},
      body: JSON.stringify({force: !!force}),
    }).then(function(r) {
      if (r.ok) {
        // The daemon retires behind this: the page's streams drop, and MASS
        // comes back on the new build a moment later. In the desktop app the
        // daemon also pushes update-restarting to the window process, which
        // calls this same function — so the notice reads identically whether it
        // was raised here or from there.
        window.massUpdateRestarting(tag);
        return;
      }
      return r.text().then(function(t) {
        installBtn.disabled = false;
        window.massAlert(window.massErrorText(t) || ('Update failed (HTTP ' + r.status + ').'),
          {title: 'Update Failed', variant: 'danger'});
      });
    }, function() {
      installBtn.disabled = false;
      window.massAlert("Couldn't start the update: MASS isn't responding.",
        {title: 'Update Failed', variant: 'danger'});
    });
  }

  function apply() {
    if (!tag) return;
    if (incompatible === 0) {
      post(false);
      return;
    }
    var dlg = document.getElementById('mass-confirm-update-dialog');
    if (!dlg) {
      // No dialog to confirm with: refuse rather than force silently.
      window.massAlert('Updating to ' + tag + ' would strand ' + workersPhrase(incompatible) +
        '. Upgrade the workers first, or run `mass update --apply --force`.',
        {title: 'Incompatible Workers', variant: 'danger'});
      return;
    }
    document.getElementById('mass-confirm-update-text').textContent =
      'Updating to ' + tag + ' would strand ' + workersPhrase(incompatible) +
      '. A stranded worker is rejected when it reconnects, exits, and stays down until it is upgraded.';
    dlg.show();
  }

  checkBtn.addEventListener('click', function() { check(); });
  installBtn.addEventListener('click', apply);

  var confirmDlg = document.getElementById('mass-confirm-update-dialog');
  if (confirmDlg) {
    document.getElementById('mass-confirm-update-cancel')
      .addEventListener('click', function() { confirmDlg.hide(); });
    document.getElementById('mass-confirm-update-ok')
      .addEventListener('click', function() { confirmDlg.hide(); post(true); });
  }

  hydrate();
})();
